src/inventory_ui.cpp

```diff
@@ -2923,55 +2923,70 @@ void inventory_examiner::action_examine( const item_location sitem, const bool f
     if( !fit_to_window ) {
         inventory_selector::action_examine( sitem );
     } else {
-        // Code below pulled from the action_examine function in advanced_inv.cpp
         std::vector<iteminfo> vThisItem;
         std::vector<iteminfo> vDummy;
 
         sitem->info( true, vThisItem );
 
         item_info_data data( sitem->tname(), sitem->type_name(), vThisItem, vDummy, examine_window_scroll );
-        //Tries to create a window covering the right half of the inventory screen
         data.without_getch = true;
 
+        draw_item_info( w_examine, data );
+    }
+}
+
+item_location inventory_examiner::execute()
+{
+    shared_ptr_fast<ui_adaptor> ui = create_or_get_ui_adaptor();
+
+    ui_adaptor ui_examine;
+
+    ui_examine.on_screen_resize( [&]( ui_adaptor & ui_examine ) {
         // NOTE: This assumes that the first column (currently own_inv_column) is used for the item list:
         int inv_column_width = get_all_columns().front()->get_width();
         /* NOTE: By flagging force_max_window_size as true, this forces the inventory_selector window to be
          * TERMX x TERMY , allowing us to assume the window starts at 0,0 and is TERMX by TERMY.  If this is
          * changed in inventory_selector::prepare_layout, then everything breaks.  This obviously isn't a
          * great solution */
-
         int border_width = 1; //Assumed, since the value in inventory_selector is private, but reasonable
         int width = TERMX - 2 * border_width  - inv_column_width;
         int height = TERMY  - get_header_height() - 1;
         point start_position = point( ( inv_column_width + border_width + 1 ),
                                       get_header_height() + 1 );
 
-        draw_item_info( catacurses::newwin( height, width, start_position ), data );
-    }
-}
-
-item_location inventory_examiner::execute()
-{
-    shared_ptr_fast<ui_adaptor> ui = create_or_get_ui_adaptor();
-    while( true ) {
-        ui_manager::redraw();
+        w_examine = catacurses::newwin( height, width, start_position );
+        ui_examine.position_from_window( w_examine );
+    } );
+    ui_examine.mark_resize();
 
+    ui_examine.on_redraw( [&]( const ui_adaptor & ) {
         const inventory_entry &selected = get_active_column().get_selected();
         if( selected ) {
-            const item_location &sitem = selected.any_item();
-            action_examine( sitem, true );
+            if( selected_item != selected.any_item() ) {
+                //A new item has been selected, reset scrolling
+                examine_window_scroll = 0;
+                selected_item = selected.any_item();
+            }
+            action_examine( selected_item, true );
         }
+    } );
+
+    while( true ) {
+        ui_manager::redraw();
 
         const inventory_input input = get_input();
 
+        //Basically anything the player can do at this point will invalidate the examine screen
+        ui_examine.invalidate_ui();
```

Comment:
`ui_manager::redraw()` always invalidates the topmost `ui_adaptor` before redrawing. `ui_manager::redraw_invalidated` can be used if you want to explicitly mark `ui_adaptor` for redrawing.

Reply:
Is there a specific sequence you recommend for this?  My testing for this was a fun cycle of "the main window draws, but it's not calling on_redraw for ui_examine"->"ui_examine redraws, but the main window doesn't anymore"->"The main window is back, but ui_examine only shows up after user input"->etc.
I remember something going wrong when I didn't invalidate ui_examine, but I'd have to go back and test.

Reply:
Never mind.  It turns out that line was a holdover from a previous iteration where ui_examine was created before calling create_or_get_ui_adaptor.  I've removed that line, but left myself a comment at the redraw() call so I remember the thought process.